### DISCUSSION TITLE
Can not upload File

### DISCUSSION THREAD

**Discussion by meideros** [adonisjs/core#3947] [Help]
I get this.adapter.config.endpoint is not a function* like error  when i  use 

moveToDisk() fonction

**Comment by meideros** [comment]
this is my stack trace 

"stack":  "TypeError: this.adapter.config.endpoint is not a function\n    at S3Driver.getUrl (/node_modules/@adonisjs/drive-s3/build/src/Drivers/S3.js:220:66)\n    at File.moveToDisk (/node_modules/@adonisjs/bodyparser/build/src/Multipart/File.js:177:44)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at Attachment.save

**Comment by meideros** [comment]
@thetutlage  @RomainLanz  it seem like s3 driver require S3_ENDPOINT instead of let it optional

**Reply by thetutlage** [reply]
Please share the code you are writing. The information you have provided is not enough to give any answer

**Reply by meideros** [reply]
const requestBody = await ctx.request.validate(StoreAndUpdatePostValidator)
    await requestBody.image.moveToDisk(
      './uploadedFiles/blogImages/',
      {
        name: `${requestBody?.image?.clientName}`,
      },
      's3'
    )
    const blogPostImageFileName = requestBody.image.fileName
    const blogPost = await BlogPostsRepository.storeBlogPost({
      ...requestBody,
      blogPostImageFileName: blogPostImageFileName,
    })

**Reply by meideros** [reply]
I alse try  
 const res = await request
      .file("image")
      ?.moveToDisk("uploadedFiles/blogImages", { visibility: "public" }, "s3");

**Reply by meideros** [reply]
@thetutlage I discover the attachment  lite package but i get the same error 

the code is 
const newPost = new Post();
const postData = await request.validate(CreatePostValidator);
 newPost.title = postData.title;
 newPost.content= postData.content;
    newPost.cover = Attachment.fromFile(postData.cover);
newPost.save()

**Reply by meideros** [reply]
Can anyone help me?

**Reply by somteacodes** [reply]
Have you included the S3_ENDPOINT in your env? 
if not, you can get the S3_ENDPOINT from here
[https://docs.aws.amazon.com/general/latest/gr/s3.html](url)
example: if regoin is us-west-2, then endpoint would be https://s3.us-west-2.amazonaws.com

**Comment by omer-genc** [comment]
Hello

## Reason

/node_modules/@adonisjs/drive-s3/build/src/Drivers/s3.js
```js
     
    // Returns URL to a given path
     
    async getUrl(location) {
        /**
         * Use the CDN URL if defined
         */
        if (this.config.cdnUrl) {
            return `${this.config.cdnUrl}/${location}`;
        }
        const href = (0, url_1.format)(await this.adapter.config.endpoint()); // this line
        if (href.startsWith('https://s3.amazonaws')) {
            return `https://${this.config.bucket}.s3.amazonaws.com/${location}`;
        }
        return `${href}/${this.config.bucket}/${location}`;
    }

```

## How can you fix

/config/drive.ts
```js
export default driveConfig({
  ...
    s3: {
      ....
      cdnUrl: Env.get('S3_CDN_URL'), // add this line S3_CDN_URL = https://bucketName.region.amazonaws.com
    },
  },
})

```
